casper/src/main/scala/coop/rchain/casper/engine/BlockReceiver.scala

```diff
@@ -0,0 +1,86 @@
+package coop.rchain.casper.engine
+
+import cats.effect.Sync
+import cats.effect.concurrent.Ref
+import cats.syntax.all._
+import coop.rchain.blockstorage.casperbuffer.CasperBufferStorage
+import coop.rchain.blockstorage.util.DoublyLinkedDag
+import coop.rchain.casper.protocol.BlockMessage
+import coop.rchain.comm.PeerNode
+import coop.rchain.models.BlockHash.BlockHash
+import coop.rchain.models.BlockMetadata
+import coop.rchain.shared.Log
+import coop.rchain.store.KeyValueTypedStore
+
+sealed trait RecvStatus
+case object Receiving   extends RecvStatus // Block receiving started
+case object Received    extends RecvStatus // Block fully received
+case object Unvalidated extends RecvStatus // Block received but not validated (put in BlockStore)
+case object Validated   extends RecvStatus // Block received and validated (ready to put in BlockMetadataStore/DAG)
+case object Invalid     extends RecvStatus // Block received but invalid
```

Comment:
I have some questions related to the scope of BlockReceiver and these statuses

Does Receiving here means that node became aware of some new hash and tries to retrieve full block from the network/disk/torrent/etc? It looks to me so, so maybe better to call this `Wanted`?

What is the difference between `Received` and `Unvalidated`? Why to distinguish between Validated and Invalid here?

Reply:
> Does Receiving here means that node became aware of some new hash and tries to retrieve full block from the network/disk/torrent/etc? It looks to me so, so maybe better to call this Wanted?

Yes, make sense. Maybe better `Waited`?

> What is the difference between Received and Unvalidated?

Block `Received` after getting `BlockMessage` message in `Running` and `Unvalidated` when it stored in `BlockStore` after some initial validation (format, signature, ...).

> Why to distinguish between Validated and Invalid here?

A block is considered `Validated` when it is removed from the `CasperBuffer`. The `Invalid` status is for invalid blocks, but it is not currently in use, so it can be removed.